### PR TITLE
[FLINK-11071][core] add support for dynamic proxy classes resolution …

### PR DESCRIPTION
**Pull request by eaglewatcherwb**
…in job defination

Change-Id: I5a7d9c489a15a2dcc0cf895e68a1d11ca57f68bc

## What is the purpose of the change

add support for dynamic proxy classes resolution in job defination

## Brief change log

  - Add custom implementation of `resolveProxyClass` in `InstantiationUtil`, which could use the costumed class loader
  - Add new unit test to verify the fix

## Verifying this change

This change added tests and can be verified as follows:

  - Without this fix, a java.lang.ClassNotFoundException will be thrown when running the added unit test.
  - With this fix, the added unit test could run properly

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency):  no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers:  no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/socket/SocketWindowWordCount.java [hunk 1]
```diff
@@ -65,33 +70,47 @@ public static void main(String[] args) throws Exception {
 		DataStream<String> text = env.socketTextStream(hostname, port, "\n");
 
 		// parse the data, group it, window it, and aggregate the counts
-		DataStream<WordWithCount> windowCounts = text
-
-				.flatMap(new FlatMapFunction<String, WordWithCount>() {
+		DataStream<WordWithCount> windowCounts = text.flatMap(serializableProxy(
```
Comment (OlegZhukov) [inline flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/socket/SocketWindowWordCount.java]: I don't think the example should be changed. Using proxies in operators is a pretty advanced use case, while the example should focus on basic usage.

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java [hunk 2]
```diff
@@ -91,6 +92,40 @@ public ClassLoaderObjectInputStream(InputStream in, ClassLoader classLoader) thr
 			return super.resolveClass(desc);
 		}
 
+		@Override
+		protected Class<?> resolveProxyClass(String[] interfaces) throws IOException, ClassNotFoundException {
```
Comment (OlegZhukov) [inline flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java]: I'm wondering if a test case can be written for this. Is there a test case for correctly resolving classes in a basic (non-proxy case)?

File: flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/socket/SocketWindowWordCount.java (hunk 1 above)
Comment (eaglewatcherwb) [reply flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/socket/SocketWindowWordCount.java]: Good comment, original case should be retained. I will add another case.

File: flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java (hunk 2 above)
Comment (eaglewatcherwb) [reply flink-core/src/main/java/org/apache/flink/util/InstantiationUtil.java]: I read the code, it seems that there is no test case for this. I will try to add a test case.

**Comment by eaglewatcherwb** [comment]
@OlegZhukov  for the `resolveProxyClass(interfaces)` calling sequence, I think `super.resolveProxyClass(interfaces)` should be called last, since there is a user defined class loader, which may be user really want to use rather than the default one.  And in `resolveClass`,  `super.resolveClass` is also called last, should we be consistent with this?

**Comment by dawidwys** [comment]
Thanks for the contribution @eaglewatcherwb .

I don't think we should introduce incompatibilities with that change. Having two different versions of an interface is a problem in the first place. If it occurs the solution would be just to simple leave only one version of the proxy. 

@eaglewatcherwb I don't think we should change `SocketWindowWordCount` example at all. I would prefer to see a proper unit test for those changes. You can check `org.apache.flink.runtime.classloading.ClassLoaderTest#testMessageDecodingWithUnavailableClass` for inspiration.

**Comment by eaglewatcherwb** [comment]
> Thanks for the contribution @eaglewatcherwb .
> 
> I don't think we should introduce incompatibilities with that change. Having two different versions of an interface is a problem in the first place. If it occurs the solution would be just to simple leave only one version of the proxy.
> 
> @eaglewatcherwb I don't think we should change `SocketWindowWordCount` example at all. I would prefer to see a proper unit test for those changes. You can check `org.apache.flink.runtime.classloading.ClassLoaderTest#testMessageDecodingWithUnavailableClass` for inspiration.

@dawidwys Thanks for the comment.
"leave only one version of the proxy." do you mean `ClassLoaderObjectInputStream.resolveProxyClass `not call `super.resolveProxyClass` and just return `null` when `classLoader==null` ?

**Comment by dawidwys** [comment]
@eaglewatcherwb No that is not what I meant. I just tried to elaborate on @OlegZhukov concern regarding backwards compatibility.

The only thing missing in this PR, from my point of view is a unit test.

**Comment by eaglewatcherwb** [comment]
> @eaglewatcherwb No that is not what I meant. I just tried to elaborate on @OlegZhukov concern regarding backwards compatibility.
> 
> The only thing missing in this PR, from my point of view is a unit test.

Thanks for the explanation. Unit test is added.

File: flink-core/src/test/java/org/apache/flink/util/InstantiationUtilTest.java [hunk 3]
```diff
@@ -46,6 +50,17 @@
  */
 public class InstantiationUtilTest extends TestLogger {
 
+	@Test
+	public void testResolveProxyClass() throws Exception {
```
Comment (dawidwys) [inline flink-core/src/test/java/org/apache/flink/util/InstantiationUtilTest.java]: This test does not check the new behavior. The proxy class is available in the same classloader as `InstantiationUtil`, which corresponds to a situation that the proxy class is available on the parent classpath in flink cluster. What you should actually test is when the proxy class is only available from user classloader. 

In `org.apache.flink.runtime.classloading.ClassLoaderTest#testMessageDecodingWithUnavailableClass` you can check how to create a new classloader with a class. Also always please make sure that your test fails without the changes that are supposed to fix the tested problem (It is not the case here, it passes without your changes).

**Comment by dawidwys** [comment]
Hi @eaglewatcherwb, I slightly improved the test. You can have a look here: https://github.com/dawidwys/flink/commit/c6f67c1a1197abc4faf6340a8e7ff3d09a8a6ba9

Basically I changed it so that:
- both interface and the invocation handler comes from the user classloader
- I used the `InstationUtil#serialize/deserialize` methods, which are the user facing ones.

I will merge it like that, thank you for your contribution!

**Comment by eaglewatcherwb** [comment]
> Hi @eaglewatcherwb, I slightly improved the test. You can have a look here: [dawidwys@c6f67c1](https://github.com/dawidwys/flink/commit/c6f67c1a1197abc4faf6340a8e7ff3d09a8a6ba9)
> 
> Basically I changed it so that:
> 
> * both interface and the invocation handler comes from the user classloader
> * I used the `InstationUtil#serialize/deserialize` methods, which are the user facing ones.
> 
> I will merge it like that, thank you for your contribution!

It's my pleasure, and nice improvement over the test.

**Pull request closed without merging**